[FSDP][9/N] Introduce `CustomPolicy`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104986
* #104969
* #104999
* #104967
* #104427

This PR adds a new `CustomPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = CustomPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root

---

After this PR, the follow-up work items for auto wrapping are:
1. Add shared parameter validation
2. (Longer-term / exploratory) Add a policy that provides a reasonable auto wrapping with "minimal" user input

Commit:
[FSDP][9/N] Introduce `LambdaWrapPolicy`

[ghstack-poisoned]

Commit:
Update on "[FSDP][9/N] Introduce `LambdaWrapPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root




[ghstack-poisoned]

Commit:
Update on "[FSDP][9/N] Introduce `LambdaWrapPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root




[ghstack-poisoned]

Commit:
Update on "[FSDP][9/N] Introduce `LambdaWrapPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root




[ghstack-poisoned]

Commit:
Update on "[FSDP][9/N] Introduce `LambdaWrapPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root

---

After this PR, the follow-up work items for auto wrapping are:
1. Add shared parameter validation
2. (Longer-term / exploratory) Add a policy that provides a reasonable auto wrapping with "minimal" user input


[ghstack-poisoned]

Commit:
Update on "[FSDP][9/N] Introduce `LambdaWrapPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root

---

After this PR, the follow-up work items for auto wrapping are:
1. Add shared parameter validation
2. (Longer-term / exploratory) Add a policy that provides a reasonable auto wrapping with "minimal" user input


[ghstack-poisoned]

rohan-varma commented on test/distributed/_composable/fully_shard/test_fully_shard_init.py:
```diff
@@ -43,12 +44,21 @@ def world_size(self) -> int:
+            if isinstance(module, nn.Sequential):
+                return True
+            elif isinstance(module, FakeSequential):
+                return {"backward_prefetch": BackwardPrefetch.BACKWARD_POST}
```
should we add a test to ensure we raise a good error when we return an invalid key? like {"foo": 1}

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -479,6 +482,96 @@ def _test_transformer_wrapping(
+
+        if use_uniform_kwargs:
+
+            def lambda_fn(module: nn.Module):
```
does LambdaPolicy still get overriden and individually wrap BatchNorms if mixed precision is configured? 

In general, what happens if my LambdaPolicy tells to wrap a BN, but mixed precision overrides this? What should be the expected behavior?

awgu replied:
Currently, our mixed precision overrides run last. I am not sure if we should special case the `LambdaWrapPolicy` to be highest priority. What do you think?

I was leaning toward just keeping our overrides as absolute for now.

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -479,6 +482,96 @@ def _test_transformer_wrapping(
+            if module is bn:
+                self.assertTrue(isinstance(module, FSDP))
+                self.assertEqual(module.sharding_strategy, bn_strategy)
+                self.assertEqual(module.backward_prefetch, bn_prefetch)
```
we never set it to backward_pre though, how does it get backward_pre configured?

awgu replied:
`BACKWARD_PRE` is default :)

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -707,12 +800,34 @@ def test_frozen_params(self):
+            self._test_frozen_params(use_orig_params, policy)
 
-    def _test_frozen_params(self, use_orig_params: bool):
+    def _test_frozen_params(self, use_orig_params: bool, policy: _WrapPolicy):
```
maybe add a test for the mixed precision overriding as well?

awgu replied:
Added!

rohan-varma commented on torch/distributed/fsdp/wrap.py:
```diff
@@ -209,6 +211,48 @@ def __repr__(self) -> str:
         return super().__repr__() + f"({self._module_classes_str})"
 
 
+class LambdaWrapPolicy(_WrapPolicy):
```
nit: add an example?

rohan-varma reviewed (commented):
Looks great overall, thanks for adding this! Just have a few questions.

ezyang commented on torch/distributed/fsdp/wrap.py:
```diff
@@ -209,6 +211,48 @@ def __repr__(self) -> str:
         return super().__repr__() + f"({self._module_classes_str})"
 
 
+class LambdaWrapPolicy(_WrapPolicy):
```
You solicited some name bikeshedding over WC, so I'll try my best here. Assuming we are keeping WrapPolicy suffix, CallbackWrapPolicy or CustomWrapPolicy both convey intent here reasonably well.

ezyang replied:
You could also try PredicateWrapPolicy. This is not a perfect match: I would call `lambda_fn` a predicate if it only returned a bool, but you return a little bit extra information sometimes.

ezyang commented on torch/distributed/fsdp/wrap.py:
```diff
@@ -209,6 +211,48 @@ def __repr__(self) -> str:
+
+    def _run_policy(
+        self,
+        root_module: nn.Module,
```
OOC, is there a way to easily determine the FQN of the module at this point? Without the root module it seems like it might be difficult, whereas, the code that calls `_run_policy` would be able to generate it without much difficulty.

awgu replied:
That is a good point. I do think we need access to the root module to be able to generate FQNs ("we" encompasses both in FSDP and for users).

speediedan replied:
In a downstream package I maintain, I've been allowing users to [provide module FQN-based complements](https://finetuning-scheduler.readthedocs.io/en/latest/advanced/fsdp_scheduled_fine_tuning.html#:~:text=users%20to%20provide-,module%20name%2Dbased%20complements,-to%20a%20given) to a given ``auto_wrap_policy`` using a custom ``_FSDPPolicy`` ([NameDrivenPolicy](https://github.com/speediedan/finetuning-scheduler/blob/3b70bbe0e00347ea00ea52c92ea47804193f0640/src/finetuning_scheduler/strategy_adapters/fsdp.py#L843C1-L866C113)). I should be able to continue offering that functionality with some minor refactoring after this PR, but I thought it might be worth mentioning in the context of this PR because: 
1. To call attention to this pattern since I've received positive user feedback regarding the utility FQN-based wrapping (to accommodate more flexible fine-tuning requirements in this case)
2. In case the team considers the functionality sufficiently useful to offer it natively at some point in the future (I'm not necessarily advocating that, just thinking aloud... either way this PR should help make such extensions cleaner)

Anyway, nice enhancement @awgu! A pleasure as always working with this thoughtfully designed API and constantly improving package.

awgu replied:
Thanks for the pointer! This is super helpful, especially learning more about the customer feedback.

I am thinking that `CustomPolicy` can stay generic, and we can push specific logic to the `lambda_fn` construction. If the user wants to use FQNs, then the user can use the same trick as your implementation where you translate FQNs to Python `id()`s and then use the `id()`s in the `lambda_fn`, which will see the `module` as argument. Like you said, after some refactoring, you should still be able to make use of `CustomPolicy`. The question is whether we should offer some API that permits FQNs directly. I think that is something that we can wait and see. For now, we can just add `CustomPolicy` that can serve as a base for other extensions.

What do you think?

speediedan replied:
> The question is whether we should offer some API that permits FQNs directly. I think that is something that we can wait and see. For now, we can just add `CustomPolicy` that can serve as a base for other extensions.

Agree completely, I think prioritizing API stability and simplicity at this juncture is the prudent approach. Adding an API that permits FQNs directly might be worth considering at some point in the future but that feature isn't sufficiently crucial to justify its inclusion at present (especially since implementation approaches like my `id()` translation one we discussed aren't too onerous).

Thanks for the quick and thoughtful response!

ezyang reviewed (approved):
Seems reasonable.

Commit:
Update on "[FSDP][9/N] Introduce `CustomPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root

---

After this PR, the follow-up work items for auto wrapping are:
1. Add shared parameter validation
2. (Longer-term / exploratory) Add a policy that provides a reasonable auto wrapping with "minimal" user input


[ghstack-poisoned]

Commit:
Update on "[FSDP][9/N] Introduce `CustomPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root

---

After this PR, the follow-up work items for auto wrapping are:
1. Add shared parameter validation
2. (Longer-term / exploratory) Add a policy that provides a reasonable auto wrapping with "minimal" user input


[ghstack-poisoned]

Commit:
Update on "[FSDP][9/N] Introduce `CustomPolicy`"


This PR adds a new `LambdaWrapPolicy` that acts like the existing `lambda_auto_wrap_policy` except it (1) leverages the new auto wrapping infrastructure and (2) allows overriding FSDP kwargs for particular instances. (1) gives it access to the validation checks (like for frozen parameters), and (2) makes it as expressive as manual wrapping. This should allow us to effectively deprecate manual wrapping if desired.

The API is as follows:
```
def lambda_fn(module: nn.Module) -> Union[bool, Dict[str, Any]]:
    ...
policy = LambdaWrapPolicy(lambda_fn)
```
The `lambda_fn` can return:
- `False` or `{}` to indicate no wrapping
- `True` to indicate wrapping while inheriting the root's FSDP kwargs
- Non-empty `dict` to indicate wrapping while overriding the specified FSDP kwargs and inheriting the rest from the root

---

After this PR, the follow-up work items for auto wrapping are:
1. Add shared parameter validation
2. (Longer-term / exploratory) Add a policy that provides a reasonable auto wrapping with "minimal" user input


[ghstack-poisoned]

awgu commented:
Unstable and unrelated:
[linux-focal-rocm5.6-py3.8 / test (default, 1, 3, linux.rocm.gpu, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/104986#15563485355)

awgu commented:
@pytorchbot merge -f "unrelated failures"
